Algorithm.Framework/Portfolio/BlackLittermanPortfolioConstructionModel.cs

```diff
@@ -0,0 +1,310 @@
+            double minimumWeight = -1.0,
+            double maximumWeight = 1.0,
+            double riskFreeRate = 0.0,
+            double tau = 0.025
+            )
+        {
+            _lookback = lookback;
+            _resolution = resolution;
+            _minimumWeight = minimumWeight;
+            _maximumWeight = maximumWeight;
+            _riskFreeRate = riskFreeRate;
+            _tau = tau;
+            _optimization = optimization;
+
+            _pendingRemoval = new List<Symbol>();
+            _symbolDataDict = new Dictionary<Symbol, ReturnsSymbolData>();
+        }
+
+        /// <summary>
+        /// Create portfolio targets from the specified insights
+        /// </summary>
+        /// <param name="algorithm">The algorithm instance</param>
+        /// <param name="insights">The insights to create portoflio targets from</param>
+        /// <returns>An enumerable of portoflio targets to be sent to the execution model</returns>
+        public override IEnumerable<IPortfolioTarget> CreateTargets(QCAlgorithmFramework algorithm, Insight[] insights)
+        {
+            var targets = new List<IPortfolioTarget>();
+
+            // remove pending
+            foreach (var symbol in _pendingRemoval)
+            {
+                targets.Add(new PortfolioTarget(symbol, 0));
+            }
+            _pendingRemoval.Clear();
+
+            var symbols = insights.Select(x => x.Symbol).Distinct();
+            if (symbols.Count() == 0)
+            {
+                return Enumerable.Empty<IPortfolioTarget>();
+            }
+
+            // Get symbols' returns
+            var returns = GetReturns(from s in symbols join sd in _symbolDataDict on s equals sd.Key select sd.Value.Returns);
+
+            // Calculate equilibrium returns
+            double[]  Π;
+            double[,] Σ;
+            GetEquilibriumReturns(returns, out Π, out Σ);
+
+            // Calculate implied equilibrium returns
+            double[,] P;
+            double[] Q;
+            if (TryGetViews(insights, out P, out Q))
+            {
+                // Create the diagonal covariance matrix of error terms from the expressed views
+                var Ω = P.Dot(Σ).DotWithTransposed(P).Multiply(_tau);
+                double[,] matrix = Matrix.Diagonal<double>(P.Dot(Σ).DotWithTransposed(P).Multiply(this._tau).Diagonal<double>());
+                if (Ω.Determinant() != 0)
+                {
+                    var invCov = Σ.Multiply(_tau).Inverse();
+                    var PTomega = P.TransposeAndDot(Ω.Inverse());
+
+                    var A = invCov.Add(PTomega.Dot(P));
+                    var B = invCov.Dot(Π).Add(PTomega.Dot(Q));
+                    Π = A.Inverse().Dot(B);
+                }
+            }
+
+            // The optimization method processes the data frame
+            double[] W;
+            _optimization.SetCovariance(Σ);
+            _optimization.SetBounds(_minimumWeight, _maximumWeight);
```

Comment:
Since these bounds don't change during runtime, they should be set in the optimization class constructor along with targetReturn.